M3T2/Tests/CategoriesTests.cs

```diff
@@ -0,0 +1,137 @@
+        await _connection.ExecuteAsync(@"
+            INSERT INTO Categories (Name)
+            VALUES (@Name1), (@Name2), (@Name3)
+        ", new { Name1 = setUpNames[0], Name2 = setUpNames[1], Name3 = setUpNames[2] });
+
+        var categories = (await _service.GetAll()).ToList();
+
+        Assert.True(setUpNames.All(name => categories.Any(category => category.Name == name)));
+    }
+
+    [Test]
+    public async Task GetByIdWorks()
+    {
+        var setUpName = "My category that I need to get by ID";
+        var setUpId = await _connection.QueryFirstAsync<int>(@"
+            INSERT INTO Categories (Name)
+            VALUES (@Name)
+            RETURNING Id
+        ", new { Name = setUpName });
+
+        var retrievedCategory = await _service.GetById(setUpId);
+
+        Assert.NotNull(retrievedCategory);
+        Assert.AreEqual(setUpName, retrievedCategory!.Name);
+    }
+
+    [Test]
+    public async Task GetByIdHandlesInvalidId()
+    {
+        var retrievedCategory = await _service.GetById(-1);
+
+        Assert.Null(retrievedCategory);
+    }
+
+    [Test]
+    public async Task CreateWorks()
+    {
+        var setUpCategory = new CategoryCreateDto { Name = "My new category" };
+
+        var newCategoryId = await _service.Create(setUpCategory);
+
+        var retrievedName = await _connection.QueryFirstOrDefaultAsync<string>(@"
+            SELECT Name
+            FROM Categories
+            WHERE Id = @Id
+        ", new { Id = newCategoryId });
+        Assert.AreEqual(setUpCategory.Name, retrievedName);
+    }
+
+    [Test]
+    public void CreateValidationWorks()
+    {
+        Assert.ThrowsAsync<BadRequestException>(() => _service.Create(new() { Name = "" }));
+        Assert.ThrowsAsync<BadRequestException>(() => _service.Create(new() { Name = new('A', CategoryEntity.NameMaxLength + 1) }));
+        Assert.ThrowsAsync<BadRequestException>(() => _service.Create(new() { Name = "Name", ParentCategoryId = -1 }));
+    }
+
+    [Test]
+    public async Task UpdateWorks()
+    {
+        var setUpId = await _connection.QueryFirstAsync<int>(@"
+            INSERT INTO Categories (Name)
+            VALUES ('Test name')
+            RETURNING Id
+        ");
+        var setUpUpdate = new CategoryUpdateDto { Name = "Test name (updated)" };
+
+        await _service.Update(setUpId, setUpUpdate);
+
+        var retrievedName = await _connection.QueryFirstOrDefaultAsync<string>(@"
+            SELECT Name
+            FROM Categories
+            WHERE Id = @Id
+        ", new { Id = setUpId });
+        Assert.AreEqual(setUpUpdate.Name, retrievedName);
+    }
+
+    [Test]
+    public void UpdateHandlesInvalidId() => Assert.ThrowsAsync<BadRequestException>(() => _service.Update(-1, new()));
+
+    [Test]
+    public async Task UpdateValidationWorks()
+    {
+        var setUpId = await _connection.QueryFirstAsync<int>(@"
+            INSERT INTO Categories (Name)
```

Comment:
How many tests would fail if you just add a new required property to the category? Can you think about some approaches to prevent these errors?

Reply:
As can be seen on the first screenshot below, 6 out of 10 existing tests start failing after adding a new required field to the `Categories` table. This is caused by how the tests themselves are written, and can be remedied by extracting code that creates set-up entities into separate functions. After that, failing tests caused by changes in the schema can be trivially fixed by changing two lines in the set-up code to account for the new field, causing the number of failed tests in this specific example to decrease to 2 out of 10 as can be seen on the second screenshot below.

Before refactoring set-up code into `DataHelper`:
![image](https://user-images.githubusercontent.com/43933025/227767783-e4030287-d06e-4bbe-9187-64d0f58d0294.png)

After refactoring set-up code into `DataHelper`:
![image](https://user-images.githubusercontent.com/43933025/227770823-04f86d44-902a-42bc-8e07-6226ac952fea.png)